[css-anchor-position-1] Transitioning from display: block to none repeatedly tears down and re-creates anchor positioned element's renderer

tuankiet65 opened the pull request:
#### c9ce4c233320ea539524620f7242b5092a47bfc4
<pre>
[css-anchor-position-1] Transitioning from display: block to none repeatedly tears down and re-creates anchor positioned element&apos;s renderer
<a href="https://rdar.apple.com/162963994">rdar://162963994</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=301056">https://bugs.webkit.org/show_bug.cgi?id=301056</a>

Reviewed by Antti Koivisto.

This condition in TreeResolver::createAnimatedElementUpdate prevents
applyAnimations from applying animated (and transitioned) property value
until anchor resolution is finished:

auto applyAnimations = [&amp;]() -&gt; std::pair&lt;std::unique_ptr&lt;RenderStyle&gt;, OptionSet&lt;AnimationImpact&gt;&gt; {
    if (hasUnresolvedAnchorPosition)
        return { WTFMove(resolvedStyle.style), OptionSet&lt;AnimationImpact&gt; { } };

If an anchor-positioned element transitions from display: (any value) to none:

* On the first interleaved resolution, resolvedStyle.style-&gt;display() == none
  (because we&apos;re transitioning to display: none). hasUnresolvedAnchorPosition is true
  (because anchor resolution always takes at least two resolutions to complete). Therefore
  applyAnimations won&apos;t apply the animated value of display (which stays the same until
  when the transition ends, then it&apos;s none). The final style (accounting for animation)
  will then have display: none. This leads to RenderTreeUpdater tearing down the renderer
  because display: none
* On the second interleaved style:
  * AnchorPositionEvaluator::evaluate marks the element as resolved, but since the element
    doesn&apos;t have a renderer, it couldn&apos;t resolve the anchor function.
  * When createAnimatedElementUpdate runs: hasUnresolvedAnchorPosition is false, so
    applyAnimations applies the animated value of display (which is the value before transition)
    This value is different than none, and currently the element doesn&apos;t have a renderer,
    so RenderTreeUpdater will create a new renderer.

This leads to two consequences:
* We needlessly tear down and create new renderer every time Document::resolveStyle is called.
  Since this is called for every animation tick, the result is we tear down and re-create new
  renderer every animation tick.
* More importantly, on the second interleaved resolution, we can&apos;t resolve anchor functions.
  So when the transition is running, the position of the anchor-positioned element is as if it
  can&apos;t resolve anchor functions. This results in the anchor-positioned element jumping
  to a position different than the position before the transition runs.

Fix this by manually patching the display property of the returned style. If (1) we&apos;re skipping
applying animation due to unresolved anchor position, (2) there will be a transition to
display: none, then we manually patch the display property to the old value. This is valid
since when transitioning from any display value to none, the animated value is the original value,
and only becomes none when the transition is complete.

Also renamed anchored-transition-display-none.html to anchored-transition-display-none-001.html
now that we have two tests for this scenario.

Test: imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-002.html

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-001-expected.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-expected.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-001-ref.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-ref.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-001.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-002-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-002.html: Added.
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::styleHasDisplayTransition):
* Source/WebCore/animation/WebAnimationUtilities.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::updateCSSTransitions const):

Canonical link: <a href="https://commits.webkit.org/302431@main">https://commits.webkit.org/302431@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7a0404ee7ff38ef85be0dbc39071fe77a5e2f75

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129080 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1339 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39916 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136460 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/80455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130951 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1270 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1216 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98278 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/80455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132027 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/981 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115627 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78924 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/905 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33742 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79739 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109351 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34240 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138933 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1132 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1100 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106817 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1184 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111963 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106644 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27148 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/927 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30487 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53647 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1205 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/64557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1031 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1077 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1129 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -791,9 +791,6 @@ ElementUpdate TreeResolver::createAnimatedElementUpdate(ResolvedStyle&& resolved
     };
 
     auto applyAnimations = [&]() -> std::pair<std::unique_ptr<RenderStyle>, OptionSet<AnimationImpact>> {
-        if (hasUnresolvedAnchorPosition)
```
nice

tuankiet65 replied:
I switched to a different approach so this no longer applies.

graouts commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -3693,6 +3693,40 @@ void RenderStyle::adjustViewTimelines()
+            [&](const CSS::Keyword::All&) {
+                if (transition.behavior() == TransitionBehavior::AllowDiscrete)
+                    return true;
+                return false;
```
```suggestion
                return transition.behavior() == TransitionBehavior::AllowDiscrete;
```

tuankiet65 replied:
I took the suggested change.

graouts commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -3693,6 +3693,40 @@ void RenderStyle::adjustViewTimelines()
+            },
+            [&](const Style::SingleTransitionProperty::SingleProperty& property) {
+                if (std::holds_alternative<CSSPropertyID>(property.value)) {
+                    if (std::get<CSSPropertyID>(property.value) == CSSPropertyDisplay) {
```
Perhaps use `std::get_if<>` here?

tuankiet65 replied:
I changed it to use `std::get_if`

anttijk commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -2502,6 +2502,8 @@ class RenderStyle final : public CanMakeCheckedPtr<RenderStyle, WTF::DefaultedOp
     inline bool insideSubmitButton() const;
     inline void setInsideSubmitButton(bool);
 
+    bool hasDisplayTransition() const;
```
It would be generally better to not pile up logic into `RenderStyle`. Perhaps this could be a standalone function somewhere in the animation code?

tuankiet65 replied:
I moved it to `Source/WebCore/animation/WebAnimationUtilities.cpp`

graouts commented on Source/WebCore/animation/WebAnimationUtilities.cpp:
```diff
@@ -390,4 +392,35 @@ AtomString animatablePropertyAsString(AnimatableCSSProperty property)
+            [&](const Style::SingleTransitionProperty::SingleProperty& property) {
+                if (auto* ptr = std::get_if<CSSPropertyID>(&property.value); ptr && *ptr == CSSPropertyDisplay)
+                    return transition.behavior() == TransitionBehavior::AllowDiscrete;
+
```
Unnecessary empty line.

graouts commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -808,8 +808,22 @@ ElementUpdate TreeResolver::createAnimatedElementUpdate(ResolvedStyle&& resolved
+            // 2. the new style has display: none and specifies a transition on display.
+            if (oldStyle && oldStyle->hasTransitions() && oldStyle->display() != DisplayType::None && styleHasDisplayTransition(*newStyle) && newStyle->display() == DisplayType::None)
+                newStyle->setDisplay(oldStyle->display());
+
```
Unnecessary empty line.

tuankiet65 committed:
[css-anchor-position-1] Transitioning from display: block to none repeatedly tears down and re-creates anchor positioned element's renderer
rdar://162963994
https://bugs.webkit.org/show_bug.cgi?id=301056

Reviewed by Antti Koivisto.

This condition in TreeResolver::createAnimatedElementUpdate prevents
applyAnimations from applying animated (and transitioned) property value
until anchor resolution is finished:

auto applyAnimations = [&]() -> std::pair<std::unique_ptr<RenderStyle>, OptionSet<AnimationImpact>> {
    if (hasUnresolvedAnchorPosition)
        return { WTFMove(resolvedStyle.style), OptionSet<AnimationImpact> { } };

If an anchor-positioned element transitions from display: (any value) to none:

* On the first interleaved resolution, resolvedStyle.style->display() == none
  (because we're transitioning to display: none). hasUnresolvedAnchorPosition is true
  (because anchor resolution always takes at least two resolutions to complete). Therefore
  applyAnimations won't apply the animated value of display (which stays the same until
  when the transition ends, then it's none). The final style (accounting for animation)
  will then have display: none. This leads to RenderTreeUpdater tearing down the renderer
  because display: none
* On the second interleaved style:
  * AnchorPositionEvaluator::evaluate marks the element as resolved, but since the element
    doesn't have a renderer, it couldn't resolve the anchor function.
  * When createAnimatedElementUpdate runs: hasUnresolvedAnchorPosition is false, so
    applyAnimations applies the animated value of display (which is the value before transition)
    This value is different than none, and currently the element doesn't have a renderer,
    so RenderTreeUpdater will create a new renderer.

This leads to two consequences:
* We needlessly tear down and create new renderer every time Document::resolveStyle is called.
  Since this is called for every animation tick, the result is we tear down and re-create new
  renderer every animation tick.
* More importantly, on the second interleaved resolution, we can't resolve anchor functions.
  So when the transition is running, the position of the anchor-positioned element is as if it
  can't resolve anchor functions. This results in the anchor-positioned element jumping
  to a position different than the position before the transition runs.

Fix this by manually patching the display property of the returned style. If (1) we're skipping
applying animation due to unresolved anchor position, (2) there will be a transition to
display: none, then we manually patch the display property to the old value. This is valid
since when transitioning from any display value to none, the animated value is the original value,
and only becomes none when the transition is complete.

Also renamed anchored-transition-display-none.html to anchored-transition-display-none-001.html
now that we have two tests for this scenario.

Test: imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-002.html

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-001-expected.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-expected.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-001-ref.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-ref.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-001.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-002-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchored-transition-display-none-002.html: Added.
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::styleHasDisplayTransition):
* Source/WebCore/animation/WebAnimationUtilities.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::updateCSSTransitions const):

Canonical link: https://commits.webkit.org/302431@main
